dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java

```diff
@@ -115,6 +118,59 @@ public Event visitOpAtomicXor(SpirvParser.OpAtomicXorContext ctx) {
     }
 
+    @Override
+    public Event visitOpAtomicSMax(SpirvParser.OpAtomicSMaxContext ctx) {
+        return visitOpAtomicExtremum(ctx.idResult(), ctx.idResultType(), ctx.pointer(),
+                ctx.memory(), ctx.semantics(), ctx.valueIdRef(), IntCmpOp.GT);
+    }
+
+    @Override
+    public Event visitOpAtomicSMin(SpirvParser.OpAtomicSMinContext ctx) {
+        return visitOpAtomicExtremum(ctx.idResult(), ctx.idResultType(), ctx.pointer(),
+                ctx.memory(), ctx.semantics(), ctx.valueIdRef(), IntCmpOp.LT);
+    }
+
+    @Override
+    public Event visitOpAtomicUMax(SpirvParser.OpAtomicUMaxContext ctx) {
+        return visitOpAtomicExtremum(ctx.idResult(), ctx.idResultType(), ctx.pointer(),
+                ctx.memory(), ctx.semantics(), ctx.valueIdRef(), IntCmpOp.UGT);
+    }
+
+    @Override
+    public Event visitOpAtomicUMin(SpirvParser.OpAtomicUMinContext ctx) {
+        return visitOpAtomicExtremum(ctx.idResult(), ctx.idResultType(), ctx.pointer(),
+                ctx.memory(), ctx.semantics(), ctx.valueIdRef(), IntCmpOp.ULT);
+    }
+
+    private Event visitOpAtomicExtremum(
+            SpirvParser.IdResultContext idCtx,
+            SpirvParser.IdResultTypeContext typeCtx,
+            SpirvParser.PointerContext ptrCtx,
+            SpirvParser.MemoryContext scopeCtx,
+            SpirvParser.SemanticsContext tagsCtx,
+            SpirvParser.ValueIdRefContext valCtx,
+            IntCmpOp kind
+    ) {
+        Register register = builder.addRegister(idCtx.getText(), typeCtx.getText());
+        Expression ptr = getPointer(ptrCtx.getText());
+        Expression value = builder.getExpression(valCtx.getText());
+        if (!(ptr.getType() instanceof IntegerType) || !(value.getType() instanceof IntegerType)) {
+            throw new ParsingException("Unexpected type at '%s' or '%s', expected integer but received '%s' and '%s'",
+                    ptrCtx.getText(), valCtx.getText(), ptr.getType(), value.getType());
+        }
+        String scope = builder.getScope(scopeCtx.getText());
+        Set<String> tags = builder.getSemantics(tagsCtx.getText());
+        tags.add(builder.getExpressionStorageClass(ptrCtx.getText()));
+        Expression originalValue = ExpressionFactory.getInstance().makeIntegerCast(ptr, TYPE_FACTORY.getArchType(), true);
+        Expression newValue = ExpressionFactory.getInstance().makeIntegerCast(value, TYPE_FACTORY.getArchType(), true);
+        Expression cmpExpr = ExpressionFactory.getInstance().makeIntCmp(originalValue, kind, newValue);
+        Expression booleanExpr = ExpressionFactory.getInstance().makeBooleanCast(cmpExpr);
+        Expression ite =  ExpressionFactory.getInstance().makeITE(booleanExpr, originalValue, newValue);
+        Expression ite_casted = ExpressionFactory.getInstance().makeIntegerCast(ite, (IntegerType) ptr.getType(), true);
+        SpirvXchg xchg = newSpirvXchg(register, ptr, ite_casted, scope, tags);
```

Comment:
I think you will need a new event SpirvRmwExtremum, and add a new compilation for it in VisitorSpirvVulkan. Then ITE should be not on the pointer but on the result register of the load. I.e. now we have

bv64 DUMMY_REG_2 = load(*&%var2, ACQ)
rmw store(*&%var2, ITE((&%var2 > bv64(-3)), &%var2, bv64(-3)), REL)

but it should be 

bv64 DUMMY_REG_2 = load(*&%var2, ACQ)
rmw store(*&%var2, ITE((DUMMY_REG_2 > bv64(-3)), DUMMY_REG_2, bv64(-3)), REL)


I would also add an end-to-end test to make sure it works correctly, for example

; @Input:  %out_result = {0, 0, 0, 0}
; @Output: forall (%out_result[0]==2 and %out_result[1]==2 and %out_result[2]==2 and %out_result[3]==2)
; @Output: forall (%out_final[0]==-3 and %out_final[1]==2 and %out_final[2]==2 and %out_final[2]==-3)
; @Config: 1, 1, 1
; SPIR-V
; Version: 1.0
; Schema: 0
               OpCapability Shader
        %ext = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %main "main"
               OpSource GLSL 450
       %void = OpTypeVoid
       %bool = OpTypeBool
       %func = OpTypeFunction %void
        %int = OpTypeInt 64 1
      %v4int = OpTypeVector %int 4
    %ptr_int = OpTypePointer Uniform %int
  %ptr_v4int = OpTypePointer Output %v4int
     %c3_neg = OpConstant %int -3
         %c0 = OpConstant %int 0
         %c1 = OpConstant %int 1
         %c2 = OpConstant %int 2
         %c3 = OpConstant %int 3
        %c64 = OpConstant %int 64
       %var0 = OpVariable %ptr_int Uniform %c2
       %var1 = OpVariable %ptr_int Uniform %c2
       %var2 = OpVariable %ptr_int Uniform %c2
       %var3 = OpVariable %ptr_int Uniform %c2
 %out_result = OpVariable %ptr_v4int Output
  %out_final = OpVariable %ptr_v4int Output
       %main = OpFunction %void None %func
      %label = OpLabel

    %result0 = OpAtomicSMin %int %var0 %c1 %c64 %c3_neg
    %result1 = OpAtomicUMin %int %var1 %c1 %c64 %c3_neg
    %result2 = OpAtomicSMax %int %var2 %c1 %c64 %c3_neg
    %result3 = OpAtomicUMax %int %var3 %c1 %c64 %c3_neg

     %final0 = OpLoad %int %var0
     %final1 = OpLoad %int %var1
     %final2 = OpLoad %int %var2
     %final3 = OpLoad %int %var3

  %ptr_res_0 = OpAccessChain %ptr_int %out_result %c0
  %ptr_res_1 = OpAccessChain %ptr_int %out_result %c1
  %ptr_res_2 = OpAccessChain %ptr_int %out_result %c2
  %ptr_res_3 = OpAccessChain %ptr_int %out_result %c3

               OpStore %ptr_res_0 %result0
               OpStore %ptr_res_1 %result1
               OpStore %ptr_res_2 %result2
               OpStore %ptr_res_3 %result3

  %ptr_fin_0 = OpAccessChain %ptr_int %out_final %c0
  %ptr_fin_1 = OpAccessChain %ptr_int %out_final %c1
  %ptr_fin_2 = OpAccessChain %ptr_int %out_final %c2
  %ptr_fin_3 = OpAccessChain %ptr_int %out_final %c3

               OpStore %ptr_fin_0 %final0
               OpStore %ptr_fin_1 %final1
               OpStore %ptr_fin_2 %final2
               OpStore %ptr_fin_3 %final3

               OpReturn
               OpFunctionEnd

Reply:
Added